test/e2e/versionskew/operator.go

```diff
@@ -0,0 +1,227 @@
+
+		Expect(f).ShouldNot(BeNil(), "framework init")
+
+		evWatcher, err = f.ClientSet.CoreV1().Events("").Watch(context.TODO(), metav1.ListOptions{})
+		Expect(err).ShouldNot(HaveOccurred(), "get events watcher")
+
+		evMutex = sync.Mutex{}
+	})
+
+	deployDriver := func(name, image string) api.PmemCSIDeployment {
+		cr := api.PmemCSIDeployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: api.DeploymentSpec{
+				DeviceMode: api.DeviceModeDirect,
+			},
+		}
+		if image != "" {
+			cr.Spec.Image = image
+		}
+		driverDep := deploy.CreateDeploymentCR(f, cr)
+		return driverDep
+	}
+
+	changeVersion := func(version string) {
+		newDeploymentName := d.Name()
+		if version != "" {
+			// deploy.Parse() expects the deployment is seperated with '-'
+			// and the version should be the 4th field.
+			newDeploymentName = d.Name() + "---" + version
+		}
+
+		// Change operator release.
+		By(fmt.Sprintf("Switching the operator to %s", newDeploymentName))
+		deployment, err := deploy.Parse(newDeploymentName)
+		framework.ExpectNoError(err, "internal error while parsing %s: %v", newDeploymentName, err)
+		deploy.EnsureDeploymentNow(f, deployment)
+	}
+
+	captureEvents := func(uuids ...types.UID) {
+		framework.Logf("Checking for events on: %s", uuids)
+
+		// Initialize events map
+		evMutex.Lock()
+		evCaptured = map[types.UID]bool{}
+		for _, uid := range uuids {
+			evCaptured[uid] = false
+		}
+		evMutex.Unlock()
+
+		timeStamp := time.Now()
+		evWait.Add(1)
+		go func() {
+			defer evWait.Done()
+			for watchEvent := range evWatcher.ResultChan() {
+				ev, ok := watchEvent.Object.(*corev1.Event)
+				if !ok || !ev.LastTimestamp.After(timeStamp) ||
+					ev.Source.Component != "pmem-csi-operator" ||
+					ev.Reason != api.EventReasonRunning {
+					continue
+				}
+				func() {
+					evMutex.Lock()
+					defer evMutex.Unlock()
+					if _, ok := evCaptured[ev.InvolvedObject.UID]; ok {
+						evCaptured[ev.InvolvedObject.UID] = true
+					}
+				}()
+			}
+		}()
+	}
+
+	stopEvenCapture := func() {
+		evWatcher.Stop()
+	}
+
+	clearEvents := func() {
+		evMutex.Lock()
+		defer evMutex.Unlock()
+		for k := range evCaptured {
+			evCaptured[k] = false
+		}
+	}
+
+	pmemImage := func(containers []corev1.Container) string {
+		for _, c := range containers {
+			if c.Name == "pmem-driver" || c.Name == "pmem-csi-operator" {
+				return c.Image
+			}
+		}
+		return ""
+	}
+
+	getOperatorImage := func() string {
+		dep, err := f.ClientSet.AppsV1().Deployments(d.Namespace).Get(context.TODO(), "pmem-csi-operator", metav1.GetOptions{})
+		Expect(err).ShouldNot(HaveOccurred(), "get operator deployment")
+
+		return pmemImage(dep.Spec.Template.Spec.Containers)
+	}
+
+	ensureDeploymentReady := func(uids []types.UID, what string) {
+		framework.Logf("Ensuring that the CRs got reconciled")
+
+		Eventually(func() bool {
+			evMutex.Lock()
+			defer evMutex.Unlock()
+			for _, uid := range uids {
+				if found, ok := evCaptured[uid]; !ok || !found {
```

Comment:
It looks like this can be simplified to `if !evCaptured[uid] { return false }`

Reply:
Yes, Done.